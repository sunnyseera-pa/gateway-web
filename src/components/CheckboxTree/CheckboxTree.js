/** @jsx jsx */
import { cx } from '@emotion/css';
import { jsx } from '@emotion/react';
import isEmpty from 'lodash/isEmpty';
import PropTypes from 'prop-types';
import React from 'react';
import ReactCheckboxTree from 'react-checkbox-tree';
import 'react-checkbox-tree/lib/react-checkbox-tree.css';
import { addCommonPropTypes } from '../../configs/propTypes';
import useCommonStyles from '../../hooks/useCommonStyles';
import Icon from '../Icon';
import LayoutBox from '../LayoutBox';
import * as styles from './CheckboxTree.styles';

const CheckboxTree = ({
	className,
	mt,
	mb,
	ml,
	mr,
	width,
	minWidth,
	maxWidth,
	icons,
	nodes,
	checkboxProps: { variant: checkboxVariant },
	...outerProps
}) => {
	const commonStyles = useCommonStyles({ mt, mb, ml, mr, width, minWidth, maxWidth });

	const formatNode = node => {
		if (isEmpty(node.children)) {
			const { children, ...rest } = node;
			return rest;
		} else if (node.children) {
			return {
				...node,
				children: node.children.map(childNodes => {
					return formatNode(childNodes);
				}),
			};
		} else {
			return node;
		}
	};

	const formattedNodes = React.useMemo(() => {
		return nodes.map(node => {
			return formatNode(node);
		});
	}, [nodes]);

	return (
		<LayoutBox {...{ mt, mb, ml, mr, width, minWidth, maxWidth }}>
			<div
				css={styles.root({
					variant: 'primary',
					hasLeafIcon: !!icons.leaf,
					hasParentIcon: !!icons.parentClose || !!icons.parentOpen,
					checkboxVariant,
				})}
				className={cx(className, 'ui-CheckboxTree')}>
				<ReactCheckboxTree nodes={formattedNodes} icons={icons} {...outerProps} />
			</div>
		</LayoutBox>
	);
};

CheckboxTree.propTypes = addCommonPropTypes({
	nodes: PropTypes.array,
	icons: PropTypes.shape({
		expandClose: PropTypes.node,
		expandOpen: PropTypes.node,
		parentClose: PropTypes.node,
		parentOpen: PropTypes.node,
		leaf: PropTypes.node,
	}),
	checkboxProps: PropTypes.shape({
		variant: PropTypes.oneOf(['primary', 'secondary']).isRequired,
	}),
});

CheckboxTree.defaultProps = {
	icons: {
		expandClose: <Icon name='chevron-right' size='lg' color='purple500' />,
		expandOpen: <Icon name='chevron-bottom' size='lg' color='purple500' />,
		parentOpen: null,
		parentClose: null,
		leaf: null,
	},
	checkboxProps: {
		variant: 'primary',
	},
};

export default CheckboxTree;
