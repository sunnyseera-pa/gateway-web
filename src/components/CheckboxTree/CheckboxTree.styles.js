import { css } from '@emotion/react';
import { mixins } from '../Checkbox/Checkbox.styles';

export const root =
    ({ hasLeafIcon, hasParentIcon, checkboxVariant: variant }) =>
    theme => {
        const {
            colors,
            base: { increment },
            components: {
                Checkbox: { height, width, variants },
            },
        } = theme;

        return css`
            .react-checkbox-tree {
                font-size: 1em;
            }
            .react-checkbox-tree > ol {
                width: 100%;
            }

            .rct-node {
                word-wrap: break-word;
            }

            ${!hasParentIcon &&
            `.rct-node-parent > .rct-text .rct-node-icon {
			display: none;
		}`}

            ${!hasLeafIcon &&
            `.rct-node-leaf > .rct-text .rct-node-icon {
				display: none;
			}`}

		label {
                display: flex;
            }

            label:hover {
                background: none;
            }

            .rct-collapse {
                display: flex;
                justify-content: center;
                align-items: center;
                flex: 0 0 ${width};
            }

            .rct-collapse:focus {
                outline: none;
            }

            .rct-collapse,
            .rct-checkbox,
            .rct-node-icon {
                margin: 0 ${increment}px 0 0;
                padding: 0;
            }

            .rct-checkbox .rct-icon {
                width: ${width};
            }

            .rct-title {
                padding: 0 0 0 ${increment}px;
            }

            .rct-text {
                padding: 0 0 ${2 * increment}px 0;
            }

            input + .rct-checkbox {
                position: relative;
                width: ${width};
                height: ${height};
            }

            label:hover .rct-icon-uncheck::before {
                background: ${colors[variants[variant].hoverBackground]};
            }

            input + .rct-checkbox .rct-icon::before {
                ${mixins.before({ colors, variants, variant, width, height })}
            }

            input + .rct-checkbox .rct-icon::after {
                ${mixins.after({ width, height })}
            }

            input:disabled + .rct-checkbox .rct-icon-uncheck::after {
                ${mixins.disabledNotChecked({ width, height })}
            }

            input + .rct-checkbox .rct-icon-check::after,
            input + .rct-checkbox .rct-icon-half-check::after {
                ${mixins.checked({ colors, variants, variant })}
            }

            input + .rct-checkbox .rct-icon-half-check::after {
                ${mixins.partial({ width, height })}
                top: 50%;
                left: 50%;
                transform: translate(-50%, -50%);
            }

            input:disabled + .rct-checkbox .rct-icon::after {
                ${mixins.disabled({ colors, variants, variant })}
            }
        `;
    };
