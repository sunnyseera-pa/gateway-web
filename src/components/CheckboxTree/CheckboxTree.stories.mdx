import { Meta, Story, Canvas } from '@storybook/addon-docs';

import CheckboxTree from './CheckboxTree';

<Meta title='Core/CheckboxTree' component={CheckboxTree} />

<Canvas>
	<Story name='Basic'>
		{props => {
			const [state, setState] = React.useState({ checked: [], expanded: ['project', 'activity', 'task'] });
			return (
				<CheckboxTree
					onCheck={checked => setState({ ...state, checked })}
					onExpand={expanded => setState({ ...state, expanded })}
					nodes={[
						{
							value: 'India',
							label: 'India',
						},
						{
							value: 'Malaysia',
							label: 'Malaysia',
						},
						{
							value: 'Pakistan',
							label: 'Pakistan',
						},
						{
							value: 'United Kingdom',
							label: 'United Kingdom',
							children: [
								{
									value: 'England',
									label: 'England',
									children: [
										{
											value: 'Eastern',
											label: 'Eastern',
											children: [
												{
													value: 'East Cambridgeshire',
													label: 'East Cambridgeshire',
												},
												{
													value: 'South Cambridgeshire',
													label: 'South Cambridgeshire',
												},
											],
										},
										{
											value: 'London',
											label: 'London',
											children: [
												{
													value: 'Lambeth',
													label: 'Lambeth',
												},
												{
													value: 'Lewisham',
													label: 'Lewisham',
												},
											],
										},
										{
											value: 'Midlands',
											label: 'Midlands',
										},
										{
											value: 'North West',
											label: 'North West',
											children: [
												{
													value: 'Salford',
													label: 'Salford',
												},
											],
										},
										{
											value: 'Nottinghamshire',
											label: 'Nottinghamshire',
										},
										{
											value: 'Oxfordshire',
											label: 'Oxfordshire',
										},
										{
											value: 'South East',
											label: 'South East',
											children: [
												{
													value: 'Manchester',
													label: 'Manchester',
												},
											],
										},
										{
											value: 'West Midlands',
											label: 'West Midlands',
											children: [
												{
													value: 'Leeds',
													label: 'Leeds',
												},
											],
										},
										{
											value: 'West Yorkshire',
											label: 'West Yorkshire',
										},
										{
											value: 'Yorkshire and The Humber',
											label: 'Yorkshire and The Humber',
										},
									],
								},
								{
									value: 'Northern Ireland',
									label: 'Northern Ireland',
								},
								{
									value: 'Scotland',
									label: 'Scotland',
								},
								{
									value: 'Wales',
									label: 'Wales',
								},
							],
						},
						{
							value: 'Worldwide',
							label: 'Worldwide',
						},
					]}
					noCascade={true}
					{...state}
				/>
			);
		}}
	</Story>
</Canvas>
