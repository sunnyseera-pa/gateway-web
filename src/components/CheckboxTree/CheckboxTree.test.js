import React from 'react';
import { render, waitFor, getByRole } from '@testing-library/react';
import CheckboxTree from '.';

const mockReactCheckboxTree = jest.fn(() => null);

const props = {
	className: 'additional-classname',
	nodes: [
		{
			value: 'India',
			label: 'India',
		},
		{
			value: 'Malaysia',
			label: 'Malaysia',
		},
	],
};

let wrapper;

jest.mock('react-checkbox-tree', () => props => {
	mockReactCheckboxTree(props);
	return <div />;
});

describe('Given the Checkbox component', () => {
	describe('When it is rendered', () => {
		beforeAll(() => {
			wrapper = render(<CheckboxTree {...props} />, {
				wrapper: Providers,
			});
		});

		it('Then matches the previous snapshot', () => {
			expect(wrapper.container).toMatchSnapshot();
		});

		it('Then has the correct className', () => {
			expect(wrapper.container.querySelector('.additional-classname')).toBeTruthy();
		});

		it('Then passes in the correct props', () => {
			const { nodes } = props;

			expect(mockReactCheckboxTree.mock.calls[0][0]).toMatchObject({
				nodes,
				icons: {
					leaf: null,
					parentClose: null,
					parentOpen: null,
				},
			});
		});
	});
});
