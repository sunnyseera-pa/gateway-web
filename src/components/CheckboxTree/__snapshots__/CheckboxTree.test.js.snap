// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Checkbox component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div
    class="additional-classname css-1sivg29 ui-CheckboxTree css-p4558p-CheckboxTree"
  >
    <div />
  </div>
</div>
`;
